react/lesson1.md

```diff
@@ -36,175 +91,431 @@ cd my-app/
-  {"category": "Electronics", "price": "$99.99", "stocked": true, "name": "iPod Touch"},
-  {"category": "Electronics", "price": "$399.99", "stocked": false, "name": "iPhone 5"},
-  {"category": "Electronics", "price": "$199.99", "stocked": true, "name": "Nexus 7"}
-];
+//CORRECT
+this.setState({
+  myName: "hello"
+});
 ```
-Steps to conceptualise the solution with `React`:
-- 1: [Break The UI Into A Component Hierarchy](https://facebook.github.io/react/img/blog/thinking-in-react-components.png)
-- 2: Build A Static Version
-- 3: Identify the State and where it should live
 
-[Exercise: Code Review](http://codepen.io/lacker/pen/vXpAgj)
+Remember the change handler we added for our difficulty setting? And the array that builds our buttons array? Time to connect those two using state!
 
-# JSX: Map Arrays to Components
+Start by defining the initial state in the constructor:
 
-When working with APIs we often have collections of things in `Arrays` which we want to map to components. This means we can loop through that data and display the customisable `dumb` component for each one.
+```javascript
+constructor() {
+  super();
+  this.state = {
+    difficulty: 10
+  };
+}
+```
 
-```jsx
+Notice the call to `super()`. We always need to call that before referencing the `this` keyword in the constructor. In this case it calls the `constructor()` of the parent class `React.Component`. `super` is not a React function, it's an ES6 keyword used in class inherritence: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super
 
-renderOrganisations() {
-    const orgData = [{ name: 'Organisation 1', borough: 'Camden'}, {name: 'Organisation 2'}];
-    return orgData.map(function renderData(organisation) {
-        return (
-            <Organisation
-                name= {organisation.name}
-                borough={organisation.borough ? organisation.borough : 'None'}
-                website= {organisation.website}
-            />
-        )
+```javascript
+  difficultyChange = (event) => {
+    this.setState({
+      difficulty: event.target.value
     });
+  };
+
+  buildGame() {
+    let buttons =  [];
+    for(var i=0; i<this.state.difficulty; i++) {
+      buttons.push(<GameButton />)
+    }
+    return buttons;
+  }
+
+  <input value={this.state.difficulty} type="number" onChange={this.difficultyChange}/>
+```
+
+Notice I'm also passing the state difficulty to our input through the `value` attribute.
+
+Now go ahead and change the difficulty level in your input, and see the number of buttons update in real time. Try to keep it under 100 to avoid your browser freezing.
+
+# Props
+
+Remember HTML attributes? 
+
+`<input value="Hervin" class="nameInput" />`
+
+Props are the inputs sent from the parent to child component via its attributes. So for our `GameButton` we could send it a prop that would specify what text it should display on the button:
+
+`<GameButton btnText="First button" />`
```

Comment:
I have a preference for not removing vowels from variable names. `btn` is obvious but I wouldn't want students to get the idea that fewer characters = better code.

Reply:
yeh for the sake of clarity here - `text` should be fine

Reply:
agree, it's already button, no need for the prefix